### PR TITLE
NPUW: PA/semi static chunked dispatch

### PR DESCRIPTION
**Pull request by parthmah14**
### Details

Extends the PA front-end from #37282 with semi-static model derivation and chunked dispatch. A dynamic dispatch from GenAI's continuous-batching pipeline can now run across fixed token-size variants, with sampled logits assembled back into the caller's order.

- Validate dispatch controls before execution: token and subsequence counts, shared block-table coverage, context bounds, and sampled-token indices. Validation and routing are pure functions in `pa_dispatch.{hpp,cpp}` with device-independent unit tests.
- Derive 1024-, 128-, and 1-token variants by cloning the PA model and fixing only `input_ids` and `position_ids`. KV cache and context dimensions remain dynamic. Models outside the flat-token contract, including embedding inputs, M-RoPE, and per-layer block tables, skip variant compilation and run 1:1 on the dynamic model.
- Execute long subsequences largest-variant-first, using a dynamic request for residual tails. Rebase each chunk's controls, share the existing KV cache pools, and scatter sampled logits into one correctly sized output tensor.
- Preserve batching when the variants would only serialize the work. A dispatch is chunked when a subsequence fills a multi-token variant; single-sequence decode uses the 1-token variant. Decode batches and short prefills run as one dynamic inference.
- Add Verbose-level dispatch summaries, tensor digests, and chunk plans. Filter `DEVICE_ID` alongside NPU properties so an explicit NPU device ID is not forwarded to the CPU compile.

This stage executes the dynamic model and all variants on CPU through the NPUW PA front-end. NPU/CPU partitioning and layout changes are separate follow-up work. The derivation and dispatcher land together so the compiled variants have an execution path.

### Validation

Recorded validation of this branch on September 1:

- All 23 PA unit tests passed: 17 dispatch-contract tests and 6 routing tests.
- Traces confirmed a 7,638-token prefill split into `7 x 1024 + 3 x 128 + 86` dynamic-tail tokens, single-sequence decode through the 1-token variant, and batched decode/short prefills through the 1:1 path.
- The four-stream CPU CB regression returned to approximately 18 ms TPOT, matching the plain-CPU reference, after removing serial per-sequence decode dispatches.

Earlier Qwen3-0.6B int4 checks through `ContinuousBatchingPipeline` covered short, >128-token, and >1024-token prompts with deterministic repeated generation. Chunked and unchunked GEMM shapes can resolve greedy near-ties differently.

### Tickets

- CVS-190137: PA front-end PoC.
- CVS-190716: semi-static model derivation.

### CODE REVIEW

**Review by dylanneve1 — APPROVED** [review]

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp [hunk 1]
```diff
@@ -4,16 +4,196 @@
...
+    const auto& lshape = outputs.front().get_partial_shape();
+    return lshape.rank().is_static() && lshape.rank().get_length() == 3 && lshape[1].is_static() &&
+           lshape[2].is_static();
+}
+
+// Compact one-line digest of a tensor for the per-dispatch I/O trace:
+// element type, shape, then the values for small tensors or min/max/mean for
+// large ones. KV cache pools pass with_data=false -- the data is the whole
+// paged cache, so only the geometry is shown.
+std::string tensor_brief(const ov::SoPtr<ov::ITensor>& tensor, bool with_data = true) {
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:94]: shouldn't kv-cache tensors count as big straight away?

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp [hunk 2]
```diff
@@ -4,16 +4,196 @@
...
+        auto lo = value_at(0), hi = lo, sum = 0.0;
+        for (std::size_t i = 0; i < n; ++i) {
+            const auto v = value_at(i);
+            lo = std::min(lo, v);
+            hi = std::max(hi, v);
+            sum += v;
+        }
+        os << " min=" << lo << " max=" << hi << " mean=" << sum / static_cast<double>(n);
+    }
+    return os.str();
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:133]: Instead it could've been an overload for a stream that would write directly into.

something like

```
struct brief {
    explicit brief(&tensor);
    std::ostream& operator<<(brief&) {
        os << ...;
        return os;
    }
};

LOG_INFO("Some message here" << brief(tensor));
````

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp [hunk 3]
```diff
@@ -4,16 +4,196 @@
...
+    std::vector<int64_t> out(n);
+    if (tensor->get_element_type() == ov::element::i32) {
+        const auto* data = tensor->data<int32_t>();
+        std::copy_n(data, n, out.begin());
+    } else if (tensor->get_element_type() == ov::element::i64) {
+        const auto* data = tensor->data<int64_t>();
+        std::copy_n(data, n, out.begin());
+    } else {
+        OPENVINO_THROW("PA: unexpected element type ", tensor->get_element_type(), " for a control tensor");
+    }
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:40]: isn't this just `tensor->data<T>`?

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp [hunk 4]
```diff
@@ -4,16 +4,196 @@
...
+bool is_chunkable_pa_model(const std::shared_ptr<ov::Model>& model) {
+    static const std::unordered_set<std::string> known = {"input_ids",
+                                                          "position_ids",
+                                                          "past_lens",
+                                                          "subsequence_begins",
+                                                          "block_indices",
+                                                          "block_indices_begins",
+                                                          "max_context_len",
+                                                          "score_aggregation_window",
+                                                          "sampled_tokens_indices"};
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:58]: Are the cases we don't support?

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp (hunk 2 above)
Comment (dmatveev) [reply src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:133]: Also this structure should go to `utils` I believe

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp [hunk 5]
```diff
@@ -4,16 +4,196 @@
...
+        std::transform(vals.begin(), vals.end(), tensor->data<int32_t>(), [](int64_t v) {
+            return static_cast<int32_t>(v);
+        });
+    } else if (type == ov::element::i64) {
+        std::copy(vals.begin(), vals.end(), tensor->data<int64_t>());
+    } else {
+        OPENVINO_THROW("PA: unexpected element type ", type, " for a control tensor");
+    }
+    return tensor;
+}
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:150]: Why do we need this cast-like convert?

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp [hunk 6]
```diff
@@ -4,16 +4,196 @@
...
+
+// [start, start + n) of a 1-D tensor, copied into a fresh same-typed tensor.
+ov::SoPtr<ov::ITensor> slice_1d(const ov::SoPtr<ov::ITensor>& src, int64_t start, int64_t n) {
+    auto out = ov::get_tensor_impl(ov::Tensor(src->get_element_type(), ov::Shape{static_cast<std::size_t>(n)}));
+    const auto esize = src->get_element_type().size();
+    const auto* base = static_cast<const uint8_t*>(src->data());
+    std::memcpy(out->data(), base + static_cast<std::size_t>(start) * esize, static_cast<std::size_t>(n) * esize);
+    return out;
+}
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:159]: What's the reason to copy? Can we use a view instead?

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp [hunk 7]
```diff
@@ -4,16 +4,196 @@
...
+
+std::shared_ptr<ov::Model> derive_pa_semi_static_model(const std::shared_ptr<ov::Model>& base_model,
+                                                       std::size_t token_dim) {
+    // Only the token-driven inputs get a fixed size (both are 1-D, checked by
+    // is_chunkable_pa_model); the context stays dynamic.
+    auto derived = base_model->clone();
+    derived->reshape({{"input_ids", ov::PartialShape{static_cast<int64_t>(token_dim)}},
+                      {"position_ids", ov::PartialShape{static_cast<int64_t>(token_dim)}}});
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:167]: Should we put an assert here that `input_ids` / `position_ids` are 1D tensor here?

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp [hunk 8]
```diff
@@ -4,16 +4,196 @@
...
+
+std::map<std::size_t, ov::SoPtr<ov::ICompiledModel>> compile_pa_semi_static_variants(
+    const std::shared_ptr<ov::Model>& base_model,
+    const std::shared_ptr<const ov::IPlugin>& plugin,
+    const std::string& device,
+    const ov::AnyMap& inner_config) {
+    std::map<std::size_t, ov::SoPtr<ov::ICompiledModel>> variants;
+    constexpr std::array<std::size_t, 3> kVariantTokenDims = {1024u, 128u, 1u};
+
+    for (const auto token_dim : kVariantTokenDims) {
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:180]:
```suggestion
    for (const auto chunk_size: kVariantTokenDims) {
```

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp [hunk 9]
```diff
@@ -4,16 +4,196 @@
...
+    return derived;
+}
+
+std::map<std::size_t, ov::SoPtr<ov::ICompiledModel>> compile_pa_semi_static_variants(
+    const std::shared_ptr<ov::Model>& base_model,
+    const std::shared_ptr<const ov::IPlugin>& plugin,
+    const std::string& device,
+    const ov::AnyMap& inner_config) {
+    std::map<std::size_t, ov::SoPtr<ov::ICompiledModel>> variants;
+    constexpr std::array<std::size_t, 3> kVariantTokenDims = {1024u, 128u, 1u};
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:178]: Make chunk sizes configurable based on the new option (to be extended) - lets file a ticket to make `NPUW_LLM_PREFILL_CHUNK_SIZE` to accept a list.

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:236]: Why do we need this change?

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp [hunk 10]
```diff
@@ -56,6 +238,31 @@ ov::npuw::PACompiledModel::PACompiledModel(const std::shared_ptr<ov::Model>& mod
     LOG_INFO("PA: compiling the dynamic PA model 1:1 on " << device);
     m_compiled_model = plugin->get_core()->compile_model(model, device, inner_config);
     OPENVINO_ASSERT(m_compiled_model != nullptr, "PACompiledModel requires a valid inner compiled model");
+
+    // The device fixes the KV cache geometry at compile time; remember the
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:242]: device?

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp [hunk 11]
```diff
@@ -56,6 +238,31 @@ ov::npuw::PACompiledModel::PACompiledModel(const std::shared_ptr<ov::Model>& mod
...
+    // block size for validating block-table coverage per dispatch. Identical
+    // across layers by construction, so the first cache input answers it.
+    for (const auto& input : m_compiled_model->inputs()) {
+        if (ov::npuw::util::is_pa_key_cache_name(input.get_any_name())) {
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:246]: are we only looking at keys? not values?

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp [hunk 12]
```diff
@@ -56,6 +238,31 @@ ov::npuw::PACompiledModel::PACompiledModel(const std::shared_ptr<ov::Model>& mod
...
+            const auto& shape = input.get_partial_shape();
+            // [num_blocks (dyn), kv_heads, block_size, head_size]
+            if (shape.rank().is_static() && shape.rank().get_length() == 4 && shape[2].is_static()) {
+                m_block_size = static_cast<std::size_t>(shape[2].get_length());
+            }
+            break;
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:252]: Shouldn't it happen under that if, where the value is assigned?

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp [hunk 13]
```diff
@@ -106,19 +313,241 @@ std::shared_ptr<ov::ISyncInferRequest> ov::npuw::PACompiledModel::create_sync_in
...
+        ChunkRequest chunk;
+        chunk.request = compiled->create_infer_request();
+        OPENVINO_ASSERT(chunk.request != nullptr, "PA chunk model requires a valid infer request");
+        for (const auto& input : compiled->inputs()) {
+            chunk.inputs.emplace(input.get_any_name(), input);
+        }
+        chunk.logits = compiled->outputs().front();
+        return chunk;
+    };
+    for (const auto& [token_dim, compiled] : variants) {
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:344]:
```suggestion
    for (const auto& [chunk_size, compiled] : variants) {
```

File: src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp [hunk 14]
```diff
@@ -106,19 +313,241 @@ std::shared_ptr<ov::ISyncInferRequest> ov::npuw::PACompiledModel::create_sync_in
...
+    LOG_VERB("PA dispatch #" << m_dispatch_idx << ": " << dispatch.sequences() << " subsequence(s), "
+                             << dispatch.tokens() << " token(s), " << dispatch.sampled_tokens_indices.size()
+                             << " sampled");
+    if (pa::variants_serve(dispatch, m_variant_token_dims)) {
+        infer_chunked(dispatch);
+        m_serve_chunked_logits = true;
+    } else {
+        m_serve_chunked_logits = false;
+        m_inner_request->infer();
+    }
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/pa_compiled_model.cpp:542]: it should look like a loop (nested loops):
- while you have tokens to process (tensor size - current position):
  - you start with the largest chunk, if you can take that much from the input, you infer it (and move the current position)
  - If you can't, you go to the smaller chunk and repeat.

If the size is 1, you dispatch to 1.

if no proper size found, today - you run a dynamic shape request.
